src/router/open-plc-utils/ether/efbu.1
.TH efbu 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
efbu - Ethernet Frame Blast Utility

.SH SYNOPSIS
.BR efbu
.RI [ options ]

.SH DESCRIPTION

.PP
Send IP broadcast frames of maximum length at maximum rate over a given interface for a given period in seconds.
The maximum frame length is ETHER_MAX_LEN as defined in <net/ethernet.h>.
The maximum rate is limited by the host processor type, the number of active processes, the interface hardware and so on.
This program does the best it can within the given execution environment.
Multi-processosr and faster hosts should see at least 85mb.
Single-processor and slower hosts will see less.

.PP
This program is for test purposes.
It has no practical application beyond flooding the network with useless traffic and annoying the network administrator.
Don't abuse it. 
.SH OPTIONS
.TP
-\fBb \fIvalue\fR
Byte value used to pad Ethernet frames. Valid values are \fB0\fR through \fB255\fR and may be specified in decimal, hexadecimal or binary. Hexadecimal is prefixed with \fB0x\fR. Binary is prefixed with \fB0b\fR. The default is \fB0xAA\fR or \fB0b10101010\fR.
.TP
-\fBd \fIaddress\fR
Replace the destination Ethernet address of each transmitted frame with the option argument. The destination address consists of 12 hex digits in upper, lower or mixed case. Octets may be seperated with colons for clarity but colons are not required.
.TP
-\fBe \fIethertype\fR
The ethertype expressed as a 16-bit hexadecimal integer. Prefix "\fB0x\fR" is optional. 
The default is \fBETH_P_802_2\fR as defined in \fBif_ether.h\fR.
.TP
.RB - h
Replace the source address of each transmitted frame with the Ethernet address of the host interface used to send the frame. This allows frames to be sent correctly from any host interface. Use option \fB-i\fR to specify an interface.
.TP
-\fBi \fIinterface\fR
The host interface used to send frames. The default is "\fBeth1\fR" because most people use "eth0" as their principle network connection; however, if the string "\fBPLC\fR" is defined in the environment then it will over-ride the program default. Any interface specified on the command line will over-ride the default.
.TP
-\fBp\fI milliseconds\fR
The pause time in milliseconds. 
The program pause this long after sending each frame. 
If \fB0\fR then the network will be flooded with frames for the duration. 
The default is \fB50\fR milliseconds which results in about 20 frames per second for the duration.
Permitted values are \fB0\fR through \fBUINT_MAX\fR as defined in <limits.h>. 
.TP
-\fBt \fImilliseconds\fR
The blast duration in milliseconds. 
The default is \fB1000\fR milliseconds seconds which is about 1 second.
Permitted values are \fB0\fR through \fBUINT_MAX\fR as defined in <limits.h>. 
.TP
.RB - v
Print transmitted frames in hex dump format on stdout as they are sent. This reduces network transmission rate, as a consequence. 
.SH ARGUMENTS
None.
.SH EXAMPLES
The following example sends an IP broadcast frame of maximum length on host interface \fBeth2\fR for a duration of \fB20\fR seconds. The destination address will be the Ethernet broadcast address because option \fB-d\fR is omitted. The source address will be the host Ethernet interface address because option \fB-h\fR is present. Frames will be send at about 20 frames per second because option \fB-p\fR is not present.
.PP
   # efbu -ieth2 -t20000 -h -e0800
.PP
The outgoing frame will look something like this when captured with wireshark.
.PP
   00000000 FF FF FF FF FF FF 00 50 04 D3 4E F2 08 00 AA AA .......P..N.....
   00000010 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000020 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000030 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000040 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000050 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000060 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000070 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000080 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000090 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000000A0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000000B0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000000C0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000000D0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000000E0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000000F0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000100 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000110 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000120 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000130 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000140 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000150 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000160 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000170 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000180 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000190 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000001A0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000001B0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000001C0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000001D0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000001E0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000001F0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000200 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000210 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000220 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000230 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000240 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000250 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000260 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000270 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000280 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000290 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000002A0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000002B0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000002C0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000002D0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000002E0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000002F0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000300 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000310 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000320 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000330 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000340 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000350 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000360 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000370 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000380 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000390 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000003A0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000003B0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000003C0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000003D0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000003E0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000003F0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000400 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000410 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000420 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000430 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000440 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000450 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000460 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000470 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000480 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000490 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000004A0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000004B0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000004C0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000004D0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000004E0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000004F0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000500 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000510 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000520 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000530 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000540 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000550 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000560 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000570 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000580 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   00000590 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000005A0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000005B0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000005C0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000005D0 AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA AA ................
   000005E0 AA AA AA AA AA AA AA AA AA AA                   ..........      
.SH SEE ALSO
.BR amp ( 1 ),
.BR efsu ( 1 ),
.BR hpav ( 1 )
.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>

'



src/router/open-plc-utils/ether/efbu.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or 
 *   without modification, are permitted (subject to the limitations 
 *   in the disclaimer below) provided that the following conditions 
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright 
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above 
 *     copyright notice, this list of conditions and the following 
 *     disclaimer in the documentation and/or other materials 
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of 
 *     its contributors may be used to endorse or promote products 
 *     derived from this software without specific prior written 
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE 
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE 
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR 
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER 
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) 
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE 
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.  
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   efbu.c - Ethernet Frame Blast Utility;
 *
 *   transmit an IP broadcast frame of maximum length at maximum rate
 *   for a given period of time in seconds;
 *
 *
 *--------------------------------------------------------------------*/


/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <time.h>
#include <sys/time.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/error.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/flags.h"
#include "../ether/channel.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/error.c"
#include "../tools/basespec.c"
#include "../tools/uintspec.c"
#include "../tools/hexencode.c"
#include "../tools/hexdump.c"
#include "../tools/todigit.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

/*====================================================================*
 *   program contants;
 *--------------------------------------------------------------------*/

#define EFBU_VLAN_TAG 0
#define EFBU_INTERFACE "PLC"
#define EFBU_ETHERTYPE ETH_P_802_2
#define EFBU_BINARY 0xAA
#define EFBU_TIMER 1000
#define EFBU_PAUSE 50

#ifndef ETHER_CRC_LEN
#define ETHER_CRC_LEN 4
#endif

#if defined (WIN32)
#define SLEEP(milliseconds) Sleep(milliseconds)
#else
#define SLEEP(milliseconds) usleep(1000*(milliseconds))
#endif

/*====================================================================*
 *
 *   void function (struct channel * channel, void * memory, unsigned extent, unsigned timer, unsigned pause);
 *
 *   transmit an IP broadcast frame of given length at maximum rate
 *   for a given period of time in seconds;
 *
 *
 *--------------------------------------------------------------------*/

static void function (struct channel * channel, void * memory, ssize_t extent, byte binary, unsigned timer, unsigned pause)

{

#if EFBU_VLAN_TAG

	struct ether_header
	{
		uint8_t ether_dhost [ETH_ALEN];
		uint8_t ether_shost [ETH_ALEN];
		uint32_t ether_vlan;
		uint16_t ether_type;
	}
	__attribute__ ((__packed__)) * frame = (struct ether_header *)(memory);

#else

	struct ether_header * frame = (struct ether_header *)(memory);

#endif

	struct timeval ts;
	struct timeval tc;
	unsigned since;
	memset (memory, binary, extent);
	if (extent > (ETHER_MAX_LEN - ETHER_CRC_LEN))
	{
		extent = ETHER_MAX_LEN - ETHER_CRC_LEN;
	}
	memcpy (frame->ether_dhost, channel->peer, sizeof (frame->ether_dhost));
	memcpy (frame->ether_shost, channel->host, sizeof (frame->ether_shost));

#if EFBU_VLAN_TAG

	frame->ether_vlan = htonl (0x8100A000);

#endif

	frame->ether_type = htons (channel->type);
	if (gettimeofday (&ts, NULL) == -1)
	{
		error (1, errno, CANT_START_TIMER);
	}
	for (since = 0; since < timer; since = (tc.tv_sec - ts.tv_sec) * 1000 + ((tc.tv_usec - ts.tv_usec) / 1000))
	{
		sendpacket (channel, memory, extent);
		if (gettimeofday (&tc, NULL) == -1)
		{
			error (1, errno, CANT_RESET_TIMER);
		}
		SLEEP (pause);
	}
	return;
}


/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"b:d:e:hi:p:t:v",
		"",
		"Ethernet Frame Blast Utility",
		"b n\tbinary byte value is (n) [" LITERAL (EFBU_BINARY) "]",
		"d x\treplace destination address with (x)",
		"e x\tethertype is (x) [" LITERAL (EFBU_ETHERTYPE) "]",
		"h\treplace source address with host address",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"t n\ttransmit for (n) milliseconds [" LITERAL (EFBU_TIMER) "]",
		"p n\tpause (n) milliseconds between frames [" LITERAL (EFBU_PAUSE) "]",
		"v\tverbose messages",
		(char const *) (0)
	};
	byte buffer [ETHER_MAX_LEN];
	byte binary = EFBU_BINARY;
	unsigned timer = EFBU_TIMER;
	unsigned pause = EFBU_PAUSE;
	signed c;
	if (getenv (EFBU_INTERFACE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (EFBU_INTERFACE));

#else

		channel.ifname = strdup (getenv (EFBU_INTERFACE));

#endif

	}
	optind = 1;
	memset (channel.peer, 0xFF, sizeof (channel.peer));
	memset (channel.host, 0xFF, sizeof (channel.host));
	channel.type = EFBU_ETHERTYPE;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'b':
			binary = (uint8_t)(uintspec (optarg, 0, 255));
			break;
		case 'd':
			_setbits (channel.flags, CHANNEL_UPDATE_TARGET);
			if (!hexencode (channel.peer, sizeof (channel.peer), optarg))
			{
				error (1, errno, "%s", optarg);
			}
			break;
		case 'e':
			channel.type = (uint16_t)(basespec (optarg, 16, sizeof (channel.type)));
			break;
		case 'h':
			_setbits (channel.flags, CHANNEL_UPDATE_SOURCE);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'p':
			pause = (unsigned)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			break;
		case 't':
			timer = (unsigned)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc)
	{
		error (1, ECANCELED, ERROR_TOOMANY);
	}
	openchannel (&channel);
	function (&channel, buffer, sizeof (buffer), binary, timer, pause);
	closechannel (&channel);
	return (0);
}

